stl/inc/flat_map

```diff
@@ -0,0 +1,509 @@
+concept _Compare_is_transparent = requires { typename _Compare::is_transparent; };
+
+template <class _Key, class _T, class _Key_compare>
+struct _Flat_value_compare {
+    struct value_compare {
+    public:
+        bool operator()(_STD pair<const _Key&, const _T&> _X, _STD pair<const _Key&, const _T&> _Y) const {
+            return _Key_compare_for_val(_X.first, _Y.first);
+        }
+
+        value_compare(_Key_compare _Comp) : _Key_compare_for_val(_Comp) {}
+
+    private:
+        _Key_compare _Key_compare_for_val;
+    };
+};
+
+template <class _Key_container, class _Mapped_container>
+struct _Flat_Container {
+    struct container {
+        _Key_container keys;
+        _Mapped_container values;
+    };
+};
+
+template <class _Ty>
+struct _NODISCARD _Clear_flat_map_scope_guard {
+    _Ty* _Clearable;
+    _Clear_flat_map_scope_guard(_Ty* _Clearable) : _Clearable(_Clearable) {}
+
+    ~_Clear_flat_map_scope_guard() {
+        if (_Clearable) {
+            _Clearable->clear();
+        }
+    }
+};
+
+// Implementation
+
+template <_STD random_access_iterator _Key_iterator_t, _STD random_access_iterator _Mapped_iterator_t>
+class _Flat_map_iterator_Impl {
+public:
+    class type {
+    public:
+        template <class Key, class T, class Compare, class KeyContainer, class MappedContainer> requires same_as<Key, typename KeyContainer::value_type> && same_as<T, typename MappedContainer::value_type>
+        friend class flat_map;
+        type() = default;
+        type(_Key_iterator_t _Key_it, _Mapped_iterator_t _Mapped_it) : _Key_it(_Key_it), _Mapped_it(_Mapped_it) {}
+        // Copy-constructible
+        type(const type&) = default;
+        // Copy-assignable
+        type& operator=(const type&) = default;
+        // Move-assignable
+        type& operator=(type&&) = default;
+        // Swappable
+        // Destructible
+        ~type() = default;
+
+        using iterator_category = input_iterator_tag;
+        using iterator_concept  = random_access_iterator_tag;
+        using difference_type   = ptrdiff_t;
+        using value_type        = _STD pair<_STD iter_value_t<_Key_iterator_t>, _STD iter_value_t<_Mapped_iterator_t>>;
+
+    private:
+        class _Arrow_Proxy: public _STD pair<_STD iter_const_reference_t<_Key_iterator_t>, _STD iter_reference_t<_Mapped_iterator_t>> {
+        public:
+            using _STD pair<_STD iter_const_reference_t<_Key_iterator_t>, _STD iter_reference_t<_Mapped_iterator_t>>::pair;
+            _Arrow_Proxy* operator->() noexcept {
+                return this;
+            }
+        };
+
+    public:
+        using pointer           = _Arrow_Proxy;
+        using reference         = _STD pair<_STD iter_const_reference_t<_Key_iterator_t>, _STD iter_reference_t<_Mapped_iterator_t>>;
+
+        reference operator*() const {
+            return reference {*_Key_it, *_Mapped_it};
+        }
+
+        pointer operator->() const {
+            return pointer {*_Key_it, *_Mapped_it};
+        }

```

Comment:
The name should be `_Arrow_proxy` (in _Ugly_snake_case).

I'm don't know why the proxy class is publicly derived from `pair`, which is still comforming but observable.

How about the following?
```suggestion
        using reference         = _STD pair<_STD iter_const_reference_t<_Key_iterator_t>, _STD iter_reference_t<_Mapped_iterator_t>>;

    private:
        class _Arrow_proxy {
        public:
            explicit _Arrow_proxy(const reference& _Rx) noexcept : _Ref{_Rx} {}

            const reference* operator->() const noexcept {
                return _STD addressof(_Ref);
            }

        private:
            reference _Ref;
        };

    public:
        using pointer = _Arrow_proxy;

        reference operator*() const {
            return reference{*_Key_it, *_Mapped_it};
        }

        pointer operator->() const {
            return pointer{**this};
        }
```

Reply:
Thanks. Your suggestion looks great.

Reply:
Fixed